osh/sh_expr_eval.py

```diff
@@ -337,29 +337,34 @@ def _StringToBigInt(self, s, blame_loc):
         bare word: variable
         quoted word: string (not done?)
         """
-        if s.startswith('0x'):
+        m = util.RegexSearch(r'^\s*0x([0-9A-Fa-f]+)\s*$', s)
```

Comment:
Hm actually I am not sure `\s` works, because libc does POSIX ERE, and `\s` is Perl/Python style

It is maybe a little surprising that no spec tests fail because of this, I guess  we are missing coverage

---

I suggest what we should do is write a helper like `NumberMatch(s, '0x([0-9...])')`  or `MatchNumber` which implicitly adds the whitespace rule, and then we use it for ALL 3 cases

And ALSO unit test it. 

I often treat regexes as code and unit test them ...  e.g.  there is a unit test for ASSIGN_ARG_RE

```
class RegexTest(unittest.TestCase):

    def testSplitAssignArg(self):
        CASES = [
            # var name, op, value
            ('s', ['s', '', '']),
            ('value', ['value', '', '']),

```

---

Or it might be possible to simply tweak some spec test cases for all 3 -- add a space, see if they fail, and then handle the spaces

Although it is tricky because this logic is somewhat hidden in a whole bunch of if statements

---

It is definitely annoying that we are using 2 different regex dialects in Oils -- Python compiled to re2c, and POSIX ERE ...  I would like it to be all one thing.

The reason I suggested the latter is because of the submatch capture, which does seem to make things easier ... in theory we can add this to our re2c stuff, which may or  may not be worth it right now


(In some future world we would use eggex for everything, although that introduces a circular dependency!)

Reply:
Good catch. `\s` is not (strictly) POSIX! But we *have* a spec test which covers the `$(( " 42 " ))` case: https://ci.oilshell.org/uuu/github-jobs/7841/ovm-tarball.wwz/_tmp/spec/osh-py/whitespace.html.

Somewhat annoyingly, this bug would have slipped past our testing because glibc (and musl libc) supports the `\s` character class shorthand with `REG_EXTENDED`... (From glibc's perspective, this is still technically POSIX compliant: "[the] interpretation of an ordinary character preceded by an unescaped <backslash> ( '\\' ) is undefined..." [1])

<details>
<summary>
Test C program validating this and links to libc sources where <code>\s</code> is parsed.
</summary>

- glibc: https://sourceware.org/git/?p=glibc.git;a=blob;f=posix/regcomp.c;h=5380d3c7b9d5139d544f06bbaaba717dc36e14e7;hb=HEAD#l1879
- musl: https://git.musl-libc.org/cgit/musl/tree/src/regex/regcomp.c#n416

```c
#include <regex.h>
#include <stdio.h>

int main() {
  regex_t pat;
  int error;
  const int errbufz = 128;
  char errbuf[errbufz];

  int cflags = REG_EXTENDED;
  int eflags = 0;

  const char *regex = "\\s+";
  const char *test = "foo \t test";

  error = regcomp(&pat, regex, cflags);
  if (error) {
    goto error;
  }

  regmatch_t match;
  error = regexec(&pat, test, 1, &match, eflags);
  if (error) {
    goto error;
  }

  printf("Found match %d:%d\n", match.rm_so, match.rm_eo);

  regfree(&pat);
  return 0;

 error:
  regerror(error, &pat, errbuf, errbufz);
  fprintf(stderr, "Error: %s\n", errbuf);
  regfree(&pat);
  return 1;
}
```
</details>

Do we test OSH with musl or other libcs? I wonder if we have any other (POSIX) portability issues.

Anyways, I like the idea that we should unit test this. Currently we are missing coverage on `$(( " 0xFF " ))` and `$(( 2#0110 ))`. I'll also update the regexs so they are POSIX compliant.

[1]: https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html#tag_09_03_05

Reply:
Oh I see, wow, I totally did not know that!  

I just know that I never use `\s` with `egrep`, which uses the libc syntax

But now I noticed it works ... I wonder when that happened ...

---

Yeah so I guess even unit testing doesn't really catch it, because our unit tests run with GNU libc -- well at least right now they do

I think the place where it might break is OS X and BSD !   There was someone interested in setting up OpenBSD CI on sourcehut, but we don't have that unfortunately ...